Add vimwiki language specification 0.1.0

chipsenkbeil opened the pull request:
Resolves #989. This is a first pass at a specification for the vimwiki language based the help document, examples from the vimwiki wiki, the pandoc implementation, and my own research of vimwiki's internals. I haven't vetted certain aspects such as how a header handles its content. Wrote up the syntax based on how I wrote my own for the [vimwiki-server](https://github.com/chipsenkbeil/vimwiki-server) crate, [vimwiki-rs](https://crates.io/crates/vimwiki).

I figure there will be a lot of discussion here, but this should be a good start. Would appreciate comments and live edits from others as seen fit. 

I also see this as not necessarily needing to line up with how certain aspects of the vimwiki plugin work today. Rather, if there's a way that we as users of vimwiki would expect syntax to behave, then that's what I would expect to be reflected here and tooling can adapt. :)

## What is not covered

- Clarity on how each element translates into HTML (I haven't gotten to that with my library yet, so that can be a revision to this spec in the future)
- Inner details about features for elements like bold also being an anchor for links (these are mentioned in the help doc, which I think is good enough for now)

## vimwiki plugin limit reached

On another note, this file seems to crash the vimwiki plugin for me about recursive depth being too great. Not sure if I hit a limitation with the vimwiki plugin or if there's something else going on, but wasn't expecting errors when loading the file.

![Screen Shot 2020-10-04 at 11 17 25 AM](https://user-images.githubusercontent.com/2481802/95020858-4f643c80-0633-11eb-9a13-79097541c1cc.png)


cc @tinmarino, @Nudin, @brennen

tinmarino commented:
Wuah ! @chipsenkbeil wonderfull job!

I'll have a look at this over the next week to add the correpondig tests.
This will point some inconsistencies and help a lot to improve both your rust server, the vim parser and the spec.

A really heavy contribution, thank you for that.

PS: what about accepting the PR fast enought to permit some foreign modification and discuss on the issue post ?

chipsenkbeil commented:
@tinmarino, it's a lot to read, so I don't expect us to have a stable first version for awhile :)

I'm fine if you'd like to fast merge and then everyone discuss and apply edits after that. We can do that or discuss in this PR and apply edits here. Either way is fine with me!

I'm on vacation this week and spent today writing this document, so going to relax for the rest of the week, although I'll still be participating in discussion for the spec since I want to make sure I have a good understanding when I go back to revise the vimwiki server.

Hope this helps kickstart standardizing vimwiki!

tinmarino commented:
My votes about the TODO
* OK: Escaping backtick, dollar, double percent, end of code block to avoid treating them as delimiter. There is an issue raised and it is easy + usefull
* OK: Inline comment anywhere in the line commenting all until the end. It may be the case already, I must check
* NO: Tag with whitespace: I prefer not because it can be complicated (5-10h)
* OK: multiple empty line are combined between paragraph to make a single 'new paragraph'.
* OK permit inline decoration in header or description
* No opinion: TODO ... if comments are removed from vimwiki before all other syntax is evaluated, we need to provide an escape mechanism, otherwise the above syntax within inline code will be removed when rendering to HTML, parsing, etc.
  * You are right. It may be not so hard in code block but I'd push that for later as it can be hard to code (3h and luck). I would say, just escape them in multiline code as we know the context for each line (and it does not change during the line).

brennen commented:
On skimming, this looks like great work.  I'm kind of slammed for the next little while, but I'll do my best to give it a more thorough reading.

chipsenkbeil commented:
I've updated the spec a little bit more, fixed a couple of typos, and also summarized the TODOs I have so far at the original post in the PR so you don't have to go hunting for them.

chipsenkbeil commented:
@tinmarino, @brennen, @Nudin checking in to see if there are any thoughts thus far. Sounds like folks were fairly busy and unable to review, but maybe this weekend will give people a little time.

I'm nearing the point where I'm going to be optimizing my parser based on the spec, so the sooner we can get to something that is agreed upon, the better for me. :)

Want to get a true 0.1.0 release of the server, including usage within vim, in the next couple of weeks.

Nudin commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+
+*Code text* is represented by the following:
+1. A backtick or grave accent (`U+0060`)
+2. Any character other than a backtick or [[#line ending|line ending]]
+3. A backtick or grave accent (`U+0060`)
```
Make naming consistent. (Ether always both names or alt-name only on the first occurrence)

chipsenkbeil replied:
You're referring to use of *backtick* versus *grave accent*? I've always referred to it as a backtick, but the formal name seems to be grave accent. I'll follow your advice and move the alt name to the first occurrence and use backtick throughout the rest.

Nudin replied:
Yes, I just wondered that 'grave accent' is mentioned in 1. and 3. but not in 2.

tinmarino replied:
I would replace the first "or grave accent" by "(alias grave accent)" and delete the second.
<= Cureently, It make think these are two different caracters.

chipsenkbeil replied:
Removed grave accent reference.

Nudin commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+==== Keyword ====
+
+Keywords are specific, case-sensitive words that have an alternative
+highlighting within vim, but serve no other special purpose.
```
Not only in highlighting within vim but also in HTML (via classes).

chipsenkbeil replied:
Good to know! I'll revise. Haven't generated HTML from vimwiki (yet), so didn't know there was anything else special. :) Think you caught a couple of other cases earlier as well (I'm working from last comment to first).

Nudin reviewed (commented):
Took me some time to review. First a big thank you for the good work! Having this specification will be an important milestone for vimwiki. I added some comments, mostly minor things.
Will comment separately on the TODOs and open questions.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+
+*Bold text* is represented by the following:
+1. An asterisk (`U+002A` aka `*`)
+2. One or more [[#link|links]], [[#keyword|keywords]], or [[#text|text]] until
```
Good to know. I originally supported this in my parser, but I removed it because it seemed like nested stylings wasn't supported. I'll revise here and add that functionality back into my parser.

Nudin replied:
It's so far is only partly supported, but I think we should allow it. (Except for recursion of course!)

tinmarino replied:
The only support missing is a multiple line nesting for html convertion, like:
```
Start bold **this is bold
this not sure *this is normal italic* and
thisnot sure** end
```
But for syntax, all nesting are ok, it was copyed from ftplugin/html.vim (a very nice implementation with region)

tinmarino replied:
Yes sorry, that is md syntax.
What I wanted to note is that for multiline typeface  __highlight__ support  exists but not __html convertion__ (but that would be cool).
For nesting, both work.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+
+*Italic text* is represented by the following:
+1. An underscore (`U+005F` aka `_`)
+2. One or more [[#link|links]], [[#keyword|keywords]], or [[#text|text]] until
+   an underscore (`U+005F` aka `_`) is encountered
```
Okay, so I should add in that it can contain additional decorated text alongside these other options, right? E.g. `[[#decorated text|decorated text]], [[#links|links]], ...`

Nudin replied:
Yes. And maybe add a note that recursion is not allowed. So no bold in a link that is in and bold text (`foo *bold [[link|*double bold*]] still bold* bar`).

chipsenkbeil replied:
Added definition for **non-repeated decorated text**.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+   an underscore (`U+005F` aka `_`) is encountered
+3. An underscore (`U+005F` aka `_`)
+
+*Bold Italic text* is represented by either of the following:
+* Form 1
+   1. Sequence `*_`
+   2. One or more [[#link|links]], [[#keyword|keywords]], or [[#text|text]] until
+      an `_*` is encountered
+   3. Sequence `_*`
+* Form 2
+   1. Sequence `_*`
+   2. One or more [[#link|links]], [[#keyword|keywords]], or [[#text|text]] until
+      an `*_` is encountered
+   3. Sequence `*_`
```
In favor of supporting nested decorated text. Got it. 👍

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+
+Decorated text supports a variety of markups across [[#link|links]],
+[[#keyword|keywords]], and [[#text|text]]. It mirrors these different HTML elements:
+* [[https://developer.mozilla.org/en-US/docs/Web/HTML/Element/strong|<strong>]]
+* [[https://developer.mozilla.org/en-US/docs/Web/HTML/Element/em|<em>]]
+* [[https://developer.mozilla.org/en-US/docs/Web/HTML/Element/s|<s>]]
+* [[https://developer.mozilla.org/en-US/docs/Web/HTML/Element/code|<code>]]
+* [[https://developer.mozilla.org/en-US/docs/Web/HTML/Element/sup|<sup>]]
+* [[https://developer.mozilla.org/en-US/docs/Web/HTML/Element/sub|<sub>]]
```
Is this because I have an HTML tag in the description? Can you inline html in vimwiki? I was expecting it to be escaped, at least within the description of a link. Otherwise, we'd need some way to escape the start/end of an HTML tag, right?

This is one of those edge cases that I didn't see mentioned in terms of how HTML is handled if written directly in vimwiki. Commonmark seems to support HTML, but only if it's the start of a line in a document as described [here](https://spec.commonmark.org/0.29/#html-blocks).

Is there a way that we'd prefer HTML to be handled? Having inline HTML would be neat, but I'd prefer to have some marker for it like we have for inline code and math.

Or have I misunderstood what is breaking HTML output?

Nudin replied:
Yes it is due to the HTML tags. I'll comment below since this is a separate missing part of the specification.

chipsenkbeil replied:
Replaced with non-tag version. e.g. `<sup>` is now `sup`

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+
+A *transclusion link* is represented by the following:
+1. The sequence `{{`
+2. A [[#link uri|link uri]]
```
Link path instead of link uri? In the vimwiki manual, you can use local or universal URLs. For link paths, they're pointers to local files.

Since you can do `{{http://vimwiki.googlecode.com/hg/images/vimwiki_logo.png}}`, wouldn't URI be the appropriate choice?

Nudin replied:
Ah, didn't remember we allowed external resource too (had mixed it up with mediawiki). Then this should be fine.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+
+A *link uri* is represented by the following:
+1. Starts with `www.`, `//`, or a [[#link uri scheme|link uri scheme]]
+   a. If starting with `www.`, we add a virtual prefix of `https://` going forward
+   b. If starting with `//`, we add a virtual prefix of `file:/` going forward
```
What is not rendered correctly? Is it that highlighting doesn't occur for these cases? Or do you have other issues occur?

I'm going off of [is_web_link](https://github.com/vimwiki/vimwiki/blob/619f04f89861c58e5a6415a4f83847752928252d/autoload/vimwiki/html.vim#L31) for the fact that `www.` is handled specially and the manual describing `//` as an absolute path to infer b.

Nudin replied:
A and b are not rendered as subitems but as paragraph in the HTML export. Haven't looked into why, might be a bug in vimwiki.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+A *named interwiki link* is represented by the following:
+1. A [[#link start seq|link start seq]]
+2. The sequence `wn.`
+3. One or more characters that are not a colon (`U+003A` aka `:`) or [[#line ending|line ending]]
```
Makes sense to me. I can't see using non-alphanumeric as being a common or desired choice, so why don't I revise to be alphanumeric only?

Nudin replied:
Sounds good to me - loosening later is always possible.

chipsenkbeil replied:
Updated to say no pipe and no link end seq, which was defined earlier.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+An *indexed interwiki link* is represented by the following:
+1. A [[#link start seq|link start seq]]
+2. The sequence `wiki`
+3. One or more digits (`0-9`), but must be `0` or higher
```
Good catch. I originally thought it was base 1 and had `must be 1 or higher`, but then realized it was zero. I'll revise.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+2. One or more characters that are not the percent sign or [[#line ending|line ending]]
+3. The percent sign (`U+0025` aka `%`)
+
+A *math block line* is a line found after a [[#beginning math block line|beginning math block line]]
+and before an [[#ending math block line|ending math block line]] and is
+comprised of zero or more characters followed by a [[#line ending|line ending]].
```
Makes sense. Would this be another definition we should include and reuse  similar to how we have common definitions like `line ending` and `whitespace characters`?

Nudin replied:
Sounds good to me.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+}}}
+
+TODO ... is there a way to escape the `$` used to mark the beginning and end
+         of an inline math element? Is `$` even a concern within an
+         inline formula? If it is, maybe an escape sequence of `\$` would
```
Good to know. One of the TODOs is about supporting escape sequences for characters we want to use outside of vimwiki-specific syntax. Check that out when you have time and let me know your thoughts!

chipsenkbeil replied:
Resolving as this will be handled in a separate issue/pr.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+A *divider row* is represented by the following:
+1. Starts at the beginning of a line
+2. Zero or more [[#whitespace character|whitespace characters]]
+3. A sequence of pairs comprised of a [[#cell boundary|cell boundary]] and
+   one or more hyphens (`U+002D` aka `-`)
```
Oh, good catch! I think I'm using the last vimwiki release on my laptop I took with me on vacation, rather than the dev version I was using early on. I'll revise.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+2. A percent sign (`U+0025` aka `%`)
+3. The sequence `date`
+4. One or more [[#whitespace character|whitespace characters]]
+5. A date string in the format `YYYY-MM-DD` where `YYYY` symbolizes a
+   four-digit year (e.g. `1990`), `MM` symbolizes a two-digit month (e.g.
+   `04`), and `DD` symbolizes a two-digit day (e.g. `23`)
```
Does vimwiki provide means to configure the format for a diary? Was going off of the help doc's examples, but didn't think to check if it was configurable. Linking out to ISO 8601 sounds fine. :)

Nudin replied:
Good point, there was at least a feature request about having this configurable…

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+elements. Relevant definitions are borrowed from
+[[https://spec.commonmark.org/0.29/#characters-and-lines|commonmark characters and lines]].
+
+A *character* in vimwiki is a valid UTF-8 code point.
```
Sounds good to me. Wrote this based on what I saw in commonmark and the fact that all strings in Rust are valid UTF-8. 😛

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+3. The sequence `{{$`
+4. An optional [[#math environment|math environment]]
+5. Zero or more [[#whitespace character|whitespace characters]]
+6. A [[#line ending|line ending]] or end of input
```
Fair enough. I'll revise. 😄

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+2. A percent sign (`U+0025` aka `%`)
+3. The sequence `title`
+4. One or more [[#whitespace character|whitespace characters]]
+5. Any sequence of [[#whitespace character|whitespace characters]] and
+   non-whitespace characters leading up to a [[#line ending|line ending]]
```
We could make a new definition such as *all until end of line* or something like that. In my parser combinator approach, I use a combinator that does exactly that in a variety of places.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+     period (`U+002E` aka `.`) or a right parenthesis (`U+0029` aka `)`) is
+     for an ordered list
+4. A [[#whitespace character|whitespace character]]
+5. An optional [[#todo attribute|todo attribute]] and additional [[#whitespace character|whitespace character]]
```
Maybe indicates that it's configurable? Or that it can be done in some other way? If it's configurable, what can be changed other than the content between `[` and `]`? The brackets themselves?

Nudin replied:
I think this is a misunderstanding. I just wanted to point out that the format is `- [ ] text` and not `- [ ]  text` (two spaces between the closing bracket and the start of the content.

chipsenkbeil replied:
@Nudin, I'm about to do a revision based on our convo + @tinmarino's comments as well. Reading back over this one, I'm not following. From what I can see, the example of `- [ ] text` you gave is the same between the good and bad. Is the expected way to have two whitespace characters between the hyphen and the square bracket? Two whitespace characters between the closing square bracket and the text? Or were you saying that there should not be two whitespace characters, only one?

The spec I wrote indicates that you have some list character(s) followed by a space before text. If you happen to have a todo item, it indicates that you need a space between the list character(s) and the todo brackets and then another space after the todo brackets and before the text.

E.g.
- `-not allowed`
- `- good to go`
- `-[ ]not allowed`
- `-[ ] not allowed`
- `- [ ]not allowed`
- `- [ ] good to go`

tinmarino replied:
@chipsenkbeil I agree with your example. I cant find mine. Do not overlook it <= it must be an error. I just wanted to say like you: one (or more) spaces before __and__ after.
Also: I think we can safely allow multiple spaces

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+of a left square bracket (`U+005B` aka `[`) and right square bracket
+(`U+005D` aka `]`). Inbetween the square brackets is a single character to
+denote the todo status and is one of the following:
+* A space (`U+0020` aka ' ') meaning 0% or incomplete
+* A period (`U+002E` aka `.`) meaning 1-33% progress
+* A lowercase o (`U+006F` aka `o`) meaning 34-66% progress
+* An uppercase O (`U+004F` aka `O`) meaning 67-99% progress
+* An uppercase X (`U+0058` aka `X`) meaning 100% or completed
+* A hyphen (`U+002D` aka `-`) meaning rejected
```
Good to know. I'm wondering if we want a section to define user-configurable elements and specify what the default would be. That way, we can still refer to those elements through the spec, even if they aren't verbatim in someone's setup.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+   of a sublist, if a new list item is a sibling, or if a new list item
+   is the sibling of a parent
+3. One of the following prefixes that determines if the type of list item:
+   * Hyphen (`U+002D` aka `-`) is for an unordered list
+   * Asterisk (`U+002A` aka `*`) is for an unordered list
```
Commented elsewhere about having a user-configurable section in the spec where we can define the default values and refer to that section throughout the rest of the spec.

E.g. We have  a section that has these definitions:

```
== User Defined ==

*Unordered list item prefixes* are user-defined characters, defaulting to Hyphen (`U+002D` aka `-`) and Asterisk (`U+002A` aka `*`). References to these prefixes refer to the default or any user-defined character.
```

Nudin replied:
Yes, something like that.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+
+A *term line* is represented by the following:
+1. Starts at the beginning of a line
+2. One or more [[#inline elements|inline elements]] before the sequence `::`
```
Tags require non-colons inbetween, right? So you can't have `:tag1:tag2::`? If someone wrote `term1:tag1:tag2:: definition`, I'd actually expect behavior based on my definition to to result in the following:

```
Definition List
    Term
        1. Text "term1"
        2. Tag "tag1"
        3. Text "tag2"
    Definition
        Text "definition"
```

I'm fine restricting tags and/or all content. In terms of feasibility, my parser currently supports inline elements (including tags) in both the term and definition. When looking up a definition, I strip away any stylings, etc. for a pure string to match.

e.g. `*bold* :term 1::: definition with [[link]]` can be looked up via "bold term1" and still supports the stylings and tags being captured.

Nudin replied:
Yes, theoretical there is no conflict – but I assume that allowing tags in the term might give us troubles at some point.

chipsenkbeil commented on doc/specification.wiki:
```diff
@@ -0,0 +1,940 @@
+A line with no characters, or a line containing only spaces (`U+0020`) or tabs
+(`U+0009`), is called a *blank line*.
+
+A *whitespace character* is a space (`U+0020`) or tab (`U+0009`).
```
You're right in that unicode supports a lot more whitespace characters. I agree about not supporting it for the near future, though. It will make things easier for tools to not need to worry about all of the variations and I can't see that many people using anything other than spaces and tabs.

From Commonmark:

> whitespace character is a space (U+0020), tab (U+0009), newline (U+000A), line tabulation (U+000B), form feed (U+000C), or carriage return (U+000D).

Nudin commented:
Some more thoughts in no particular order:
- We should not necessarily limit the specification to what vimwiki currently does, but specify how vimwiki should do and if necessary create issues for the bugs in vimwiki
- In doubt it's better to restrict stronger, since loosening specification to allow more things can always be done, but when restricting there is the risk of breaking someone's stuff.
- We should create an overview where we collect features and edge cases and compare whether they are handled correctly by: 
   - Vimwiki "functioning" (Commands and mappings; better name is welcome)
   - Vimwiki syntax highlighting
   - Vimwikis HTML export
   - Pandoc
   - vimwiki-server crate
- I agree, that we should not allow whitespaces in tags.
- I also think that we need a way of escaping, but would vote for handling this in a separate Issue + PR. We will have to warn users, since it could break existing pages and there are a lot of users using vimwiki in many different ways.
- I would like to support multiple properties for preformatted text but prefer space as delimiter, since it is what is used in HTML as well as some other markup languages (including mediawiki) and we haven't used semicolon so far for markup (better "save" it for potential future features that needs a character).
- The semantics of transclusion isn't defined in the document yet – which is fine for version 0.1. But: please let us state that the included document must be valid, balanced content. What I mean is you may not have a file A with content `*foo` and a file B that contains `{{A}} bar*` and expect that you get it rendered as **foo bar**. This is something that mediawiki allowed and it's causing massive problems forever.

chipsenkbeil commented:
>     * We should not necessarily limit the specification to what vimwiki currently does, but specify how vimwiki should do and if necessary create issues for the bugs in vimwiki

Yes, I agree wholeheartedly there. I'm still a novice vimwiki user, but I've been using it daily for the last couple of months for work-related notes (in a dairy) and a couple of global pages to remember how I accomplished certain things like setting up Let's Encrypt.

>     * In doubt it's better to restrict stronger, since loosening specification to allow more things can always be done, but when restricting there is the risk of breaking someone's stuff.

Agree here as well. I'm also fine in leveraging the zero major version part of the spec, meaning that we can go in and change things as we see fit without worrying about breaking tools, etc. for the short term while we figure out any additional edge cases.

>     * We should create an overview where we collect features and edge cases and compare whether they are handled correctly by:
>       
>       * Vimwiki "functioning" (Commands and mappings; better name is welcome)
>       * Vimwiki syntax highlighting
>       * Vimwikis HTML export
>       * Pandoc
>       * vimwiki-server crate

I had a similar thought, but didn't want to delay releasing a spec by trying to review every implementation in detail right now. That and I don't think I would have the best perspective on the vimwiki internals, as you've already caught some criteria I missed, @Nudin

>     * I also think that we need a way of escaping, but would vote for handling this in a separate Issue + PR. We will have to warn users, since it could break existing pages and there are a lot of users using vimwiki in many different ways.

My immediate thought is to have a flag in vimwiki used for compatibility, where you can enable/disable escaping and other vimwiki syntax changes prior to this spec being defined. Tooling could also support this legacy flag when parsing as well.

>     * I would like to support multiple properties for preformatted text but prefer space as delimiter, since it is what is used in HTML as well as some other markup languages (including mediawiki) and we haven't used semicolon so far for markup (better "save" it for potential future features that needs a character).

Sounds good. I'll revise.

>     * The semantics of transclusion isn't defined in the document yet – which is fine for version 0.1. But: please let us state that the included document must be valid, balanced content. What I mean is you may not have a file A with content `*foo` and a file B that contains `{{A}} bar*` and expect that you get it rendered as **foo bar**. This is something that mediawiki allowed and it's causing massive problems forever.

I see. So don't allow imported vimwiki content to then be evaluated. Instead, import it verbatim as plain text?

Nudin commented:
> I see. So don't allow imported vimwiki content to then be evaluated. Instead, import it verbatim as plain text? Or evaluate prior to importing?

An ideal parser should parse the embedding vimwiki page and the embedded vimwiki page separately to make sure both are valid vimwiki and if not throw an error or at least a warning. After that it's open to the implementation how to do the import.

Nudin commented:
One more feature (that is hidden too well in the help page) is the setting `g:vimwiki_valid_html_tags` that specifies a list of HTML-Tags that if found in vimwiki markup are not escaped but transferred to the output when converting to HTML. All other tags or uses of `<` and `>` are escaped when converting to HMTL.
The default is: `b,i,s,u,sub,sup,kbd,br,hr`

chipsenkbeil commented:
> > I see. So don't allow imported vimwiki content to then be evaluated. Instead, import it verbatim as plain text? Or evaluate prior to importing?
> 
> An ideal parser should parse the embedding vimwiki page and the embedded vimwiki page separately to make sure both are valid vimwiki and if not throw an error or at least a warning. After that it's open to the implementation how to do the import.

Okay, that makes sense and would catch a lot of unintended mistakes if someone left off a decoration or something in one file when importing it into another.

> One more feature (that is hidden too well in the help page) is the setting `g:vimwiki_valid_html_tags` that specifies a list of HTML-Tags that if found in vimwiki markup are not escaped but transferred to the output when converting to HTML. All other tags or uses of `<` and `>` are escaped when converting to HMTL.
> The default is: `b,i,s,u,sub,sup,kbd,br,hr`

That's a bit unexpected. I would have assumed either vimwiki supported injecting all HTML (by just detecting open/closing tags) or none of it, but not an arbitrary set of tags. No wonder it randomly broke in generation :D 

Do you have a recommendation of what we should use instead of verbatim `<sup>` and other tags?

Separately, is this a feature we expect parsers to support as well outside of vimwiki? Feels like a legacy feature, especially considering it has default elements as I would have thought no elements would have made a saner default.

I'd rather have something like this:

```
Some #<b>bold text</b># using explicit, inline no-escape syntax.

###
<ul>
    <li>Also a no escape syntax</li>
    <li>at the block level instead of</li>
    <li>inline like above</li>
</ul>
###
```

Nudin commented:
> That's a bit unexpected. I would have assumed either vimwiki supported injecting all HTML (by just detecting open/closing tags) or none of it, but not an arbitrary set of tags. No wonder it randomly broke in generation :D
> 
> Do you have a recommendation of what we should use instead of verbatim `<sup>` and other tags?
> 
> Separately, is this a feature we expect parsers to support as well outside of vimwiki? Feels like a legacy feature, especially considering it has default elements as I would have thought no elements would have made a saner default.

Yes this is tricky. I'm also not happy with the default value, since most of its tags have corresponding vimwiki syntax. Maybe we can change the default in future versions, the feature is however much needed since several users use vimwiki to write their websites or blogs and use this to be able to use HTML-elements without correspondence in vimwiki ([example](https://schoenitzer.de/blog/2019/Gender%20in%20Surveys.html)).
As for the standard: I think we should mention it but consider/declare it an optional feature that some implementations have.

chipsenkbeil commented:
Thanks for the reply! I've been feverishly working on improving my parser and it involved a very extensive rewrite; however, I was able to get the time to be on par with pandoc (from the original **35 minute** parse time on `specification.wiki`) while supporting looking up elements under cursor, moving elements around in a page, indexing elements across pages, etc.

In fact, when running an optimized build of the parser library, the vimwiki crate is finally beating pandoc when parsing the `specification.wiki` file:

| Name | Build | Runtime |
|-------|------|----------|
| vimwiki-rs | Debug | 2.5s |
| pandoc | Release | 0.82s |
| vimwiki-rs | Release | 0.33s |

Now that I've gotten the parser good enough (there are other optimizations that could make it faster), I'm going to work on revising this spec based on our dialog.

> Yes this is tricky. I'm also not happy with the default value, since most of its tags have corresponding vimwiki syntax. Maybe we can change the default in future versions, the feature is however much needed since several users use vimwiki to write their websites or blogs and use this to be able to use HTML-elements without correspondence in vimwiki ([example](https://schoenitzer.de/blog/2019/Gender%20in%20Surveys.html)).

My first thought would be to add a deprecation warning when generating HTML if either
a. We detect that the tag setting hasn't been altered to let folks know that if they rely on the default tags being available, they should explicitly set them in their vimrc
b. We detect some usage of those HTML tags and detect that the tag setting hasn't been altered; a little more work, but would be more specific to those users

Alternatively, adding a disclaimer somewhere could also work, but I don't know where that would go where people would read it.

> As for the standard: I think we should mention it but consider/declare it an optional feature that some implementations have.

Makes sense. I didn't even know that was supported until you mentioned it. Don't know if pandoc supports functionality like that.

What are your thoughts on adding some sort of verbatim syntax like I mentioned in my earlier comment? To me, it would be a little more future-proof as it would clearly mark an area of the document where non-vimwiki syntax is used that would go to output. I have plans with vimwiki-server to support more than just HTML as an output of vimwiki as my immediate need involves a couple of other common formats, but couldn't depend on pandoc to do for me.

tinmarino commented:
Helo partners.

I edited the first post, added some votes. We agree with @chipsenkbeil !
It would be nice to get other votes.

Estimated time: 5min on phone

chipsenkbeil commented:
Isn't that syntax for commonmark? To my knowledge, the vimwiki language
stylings only apply on the same line. Bold uses * while italic uses _.
Correct me if I'm wrong, @tinmarino

Also, @Nudin brought up nested styles for vimwiki in an earlier comment.

I still need to revise based on discussion thus far.

On Sat, Oct 24, 2020, 6:05 PM Tinmarino <notifications@github.com> wrote:

> *@tinmarino* commented on this pull request.
> ------------------------------
>
> In doc/specification.wiki
> <https://github.com/vimwiki/vimwiki/pull/1025#discussion_r511523577>:
>
> > +}}}
> +
> +===== Syntax =====
> +
> +*Decorated text* is represented by one of the following:
> +* [[#bold text|Bold text]]
> +* [[#italic text|Italic text]]
> +* [[#bold italic text|Bold italic text]]
> +* [[#strikeout text|Strikeout text]]
> +* [[#code text|Code text]]
> +* [[#superscript text|Superscript text]]
> +* [[#subscript text|Subscript text]]
> +
> +*Bold text* is represented by the following:
> +1. An asterisk (`U+002A` aka `*`)
> +2. One or more [[#link|links]], [[#keyword|keywords]], or [[#text|text]] until
>
> The only support missing is a multiple line nesting for html convertion,
> like:
>
> Start bold **this is bold
> this not sure *this is normal italic* and
> thisnot sure** end
>
> But for syntax, all nesting are ok, it was copyed from ftplugin/html.vim
> (a very nice implementation with region)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vimwiki/vimwiki/pull/1025#discussion_r511523577>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAS55CUJZWAIUZ6V6RN7YQ3SMNMVLANCNFSM4R6RLQKQ>
> .
>

chipsenkbeil commented:
I've revised based on some of the comments and marked ones as resolved that I think are finished.

I also moved comments to be inline elements as this is the path that pandoc and vimwiki-server have both gone. It makes it such that comments can be included in code without needing to escape them. Pretty much anywhere else, comments will work as intended.

We can leave the escape sequence discussion and implementation for a separate issue & PR per @Nudin's comment.

Since I wrote this spec, I also introduced the concept of "Inline Block Element", mirroring the old HTML "inline-block" CSS display value. This is currently only comprised of a term, definition, and list item, but enables providing some meaning behind elements that fit within a block element but are not necessarily an inline element. I'll see about adding that somewhere in the spec.

chipsenkbeil commented:
I've cleaned out the TODOs as most have been answered. Made #1044 to facilitate discussion around escape sequences. Made #1045 to facilitate discussion around new verbatim syntax.

If we treat this specification (version 0.1.0) as geared towards syntax only - not how certain features operate like transclusion links, bold anchors, etc. - I think it's about ready for integration. Later versions can introduce the aforementioned functionality, HTML generation, etc.

@tinmarino, @brennen, @Nudin do you have any additional comments? Make sure to double-check that I addressed everything as well. :)

chipsenkbeil commented:
Actually, found a few new cases that need to be considered:

### Tables, Preformatted Text, and Math Blocks within List Items

```
- This is some item
    {{{
    Some code within a code block
    This looks like it would be within a list item
    }}}
- This is a second item
```


### Adjusting Start of Preformatted Text & Math Blocks

Should the following be treated equally?

```
    {{{
    some
    code
    }}}

    {{$
    some math
    block
    }}$
```

```
{{{
some
code
}}}

{{$
some math
block
}}$
```

### When does a paragraph end?

By definition, it was on blank lines, but it seemed clear that other criteria could apply, which is what I highlighted in the spec.

```
This is a paragraph
that is on multiple lines
----
Above was a divider
so this is a new paragraph
```

One specific case that is interesting is handling 4+ spaces of indentation for a block quote without adding a blank line between that and a paragraph:

```
This is
a paragraph
    Is this a block quote?
    Or is this still part of paragraph?
```

If we go by the idea that a paragraph's trailing lines can have arbitrary space in front that is trimmed off, I would think the above would all be part of a paragraph. Not sure what others would expect, though.

### Empty preformatted text and math block allowed?

```
{{{
}}}

{{$
}}$
```

----

@Nudin, @tinmarino, @brennen could use some feedback here. I now have a working prototype of a plugin (https://github.com/chipsenkbeil/vimwiki-server.nvim) leveraging the vimwiki-server and was going to start moving onto my feature list beyond text objects pretty quickly, but I wanted to get the language parser as close to accurate as possible. When implementing the final integration test against pandoc's example document, I found a couple of edge cases like the above.

tinmarino commented:
As usual only my personal opinion:

> Should the following be treated equally?
(about indented code block)
Yes, exactly as you say, you remove the indentation for the inner code. If some inner code is under indented, it will go to column 1.
It is very usefull when embeding in list. to indent for the visual apsect of the wiki. But in html, it is nonsense, I mean it should be part of the list anyway, this is css stuff then.

 >     Is this a block quote?
As indented yes. I actually hokked (Perl preparser) my markdown to get this behaviour.

> Empty preformatted text and math block allowed?
why not? Does it trouble you with the implementation? Otherwise, what is inside the block is not our concern so it can be void.

chipsenkbeil commented:
> Yes, exactly as you say, you remove the indentation for the inner code. If some inner code is under indented, it will go to column 1.
> It is very usefull when embeding in list. to indent for the visual apsect of the wiki. But in html, it is nonsense, I mean it should be part of the list anyway, this is css stuff then.

I assumed this was the case from some of the examples I saw. I'll need to update vimwiki-server and the spec to support this.

> As indented yes. I actually hokked (Perl preparser) my markdown to get this behaviour.

Okay, so you would expect 4+ spaces following immediately a paragraph to be considered a blockquote:

```
Some paragraph
with text
    and this would be
    a blockquote
and another paragraph
with more text
```

That makes things easier and is how the spec defines it today.

> > Empty preformatted text and math block allowed?
> 
> why not? Does it trouble you with the implementation? Otherwise, what is inside the block is not our concern so it can be void.

Was more confirming this was the expectation. I've already updated vimwiki-server to handle it this way. This will need to be updated in the spec.

chipsenkbeil commented:
@tinmarino @Nudin do we just want to merge this into the repo at this point? There are only a couple of minor additions that need to be made that I can do this coming weekend in a separate PR.

tinmarino commented:
@chipsenkbeil yes we do want to merge at some point.
Anytime, on your command.

chipsenkbeil commented:
Let's merge now. I can open PRs for any small changes and it will be easier
to have a dialog that way.

On Tue, Dec 22, 2020, 4:38 PM Tinmarino <notifications@github.com> wrote:

> @chipsenkbeil <https://github.com/chipsenkbeil> yes we do want to merge
> at some point.
> Anytime, on your command.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vimwiki/vimwiki/pull/1025#issuecomment-749813467>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAS55CVU25YLDYSASKZBWLTSWENVLANCNFSM4R6RLQKQ>
> .
>

tinmarino commented:
Wondefull adition.
Thank you @chipsenkbeil
